### ISSUE TITLE
CSPRNG question

### ISSUE THREAD

**Issue by gazbor** [paypal/seifnode#3]
Do you have any good reason to use a modified ISAAC implementation instead of AES as a CSPRNG? I'm working on a project with similar requirements that the SEIF protocol has and trying to learn from your solution but it seems unnecessarily complicated. I'm not the crypto expert in the team, but I can tell you that our approach is based on AES and works the following way. We are gathering entropy from different sources (camera, user interaction, and the OS dev/random and its windows equivalent) just as you for seeding the RNG time to time. The random number generator itself based on the AES block cipher and the random seed (which is a 256-bit number) is used directly as a key. In addition to the seed the RNG has another state which is basically a 128-bit counter. The random number generation happens through encrypting the next value of the counter. It does not use any mode of operation and IV, it directly encrypts the 128-bit counter as a block so the result is a also a 128-bit value whose bytes is XORed together to get a single byte. (Our generator produces one byte on every iteration) As I mentioned the RNG is reseeded periodically with truely random numbers. If the seed is random, the security of this generator is entirely based on the security of AES. (At least we think so)
So what do you think of our method and why have you chosen ISAAC and a more complex way to generate pseudo random numbers?

**Comment by thomwiggers** [comment]
Why not just use the operating system defined RNG? See also #1

**Comment by aashish-sheshadri** [comment]
@gazbor Your approach is quite good. I believe what you are describing is very similar to AES in CTR mode adapted to generate random bytes. Here is a NIST recommendation for using the same (http://csrc.nist.gov/publications/nistpubs/800-90A/SP800-90A.pdf). Fortuna which is a very good CPRNG too makes use of AES in CTR mode. Hence, choosing a pre-existing approach might be the way to go!

To that end, there are multiple options when it comes to choosing a CPRNG. We decided to choose ISAAC. We did little to no modification to how ISAAC works. We only add an additional stage on the output end. We generate excess bytes to be hashed as our output -- enabling entropy redistribution.

Of course our method has many parallels, it is out here to make sure we as a community converge to a reasonably secure solution. We are open to moving to a better algorithm if it makes a significant leap in its security guarantee.
